### PR TITLE
Use setLayerType for animations

### PR DESCRIPTION
**Pull request by ulyssesdotcodes**
This helps animations run smoothly when there's a lot going on.

### CODE REVIEW

**Comment by dlew** [comment]
Hmm, doesn't the expand have a bunch of moving parts? Those will keep getting invalidated on each frame.

Try the dev option "show hardware layer updates" and check that it isn't flashing green the whole time - it might be slightly better, but not as good as it could be, if that's the case.

**Comment by ulyssesdotcodes** [comment]
@dlew How will they keep getting invalidated?

Right now when I show hardware layer updates the buttons stay green throughout the animation.

**Comment by dlew** [comment]
Staying green is a bad thing. That means the hardware layer is constantly being re-rendered.

What happens when you use setLayerType is that it caches the entire displaylist at once. A displaylist is a tree-like structure of rendering calls. Without setLayerType it has to go and do every render call, but with setLayerType it can skip that...

...Except that it's smart enough to know when the displaylist has been invalidated. That can happen anytime any of the children change. So, for example, if the buttons are flying out at different speeds, it'll invalidate the topmost displaylist constantly.

In other words, if the entire displaylist is changing at once (like scaling, rotating, moving, etc.) then setLayerType is great. Otherwise it'll be somewhat of a burden.

What would need to be done in this case is that each button would have to have its HW layer set individually.

**Comment by ulyssesdotcodes** [comment]
Ahh got it. It caches the drawing of a view into a texture and renders that instead of the entire thing.

**Comment by ulyssesdotcodes** [comment]
@dlew updated. The whole fab still flashes green, but it profiles as using much less gpu and it feels smoother.

**Comment by dlew** [comment]
This problem  inspired me to finish off that blog post I'd been planning: http://blog.danlew.net/2015/10/20/using-hardware-layers-to-improve-animation-performance/

I'm not sure I'm comfortable pushing this out with an animation where it's constantly flashing green. Shouldn't it be very possible to avoid that problem with multiple layers?

**Comment by ulyssesdotcodes** [comment]
@dlew Ah, my bad. I was using an old version while testing. This one flashes once and then no more.

**Comment by dlew** [comment]
Then this is great!

We should see if the base library will absorb this change, would be a nice boost for everyone.

**Comment by ulyssesdotcodes** [comment]
@dlew Yeah, I was going to do that.
